### PR TITLE
fix: skip points outside of archive retention on first pass

### PR DESCRIPTION
**Pull request by jesusvazquez**
We have realized that during first pass and processing of the archive, we are counting in points outside of the retention and this can mess up some calculations. So we are fixing the code to only include points within the expected archive boundaries min timestamp and max timestamp. 

Previous we kept points in a lower archive if that timestamp did not exist in a higher-resolution archive. This was incorrect behavior because the lower archive might be a SUM aggregation, and that results in spike points amidst the original high-res points. Instead, we only take points if there is no higher-resolution archive covering that time span. Similarly, we do not keep any points in a high res archive if those points are covered by a lower-res archive. In effect, we always know which points are valid based on the boundaries between the archives:

```
   MaxT     MinT
0: [XXXXXXXXXX] 1d, 1m
1: [           XXXXXXXXXXXXXXXXX] 1w, 10m
2: [                            XXXXXXXXXXXXX]  1y, 60m
```

Based on this new approach, we can greatly reduce the amount of logic in the ReadSamples function and make it much faster and more space-efficient. We also discovered new edge cases and added tests for those.

### CODE REVIEW

File: pkg/graphite/convert/whisperconverter/whisper.go [hunk 1]
```diff
@@ -54,53 +54,60 @@ func ReadPoints(w Archive, name string) ([]whisper.Point, error) {
...
-				return nil, fmt.Errorf("duplicate timestamp at same precision in archive %s: %d", name, allPoints[i].Timestamp)
+
+		for _, p := range archivePoints {
+			if p.Timestamp < minArchiveTs {
+				continue
+			}
+			// If we have already seen a point with the same timestamp it means
+			// we already have a point from an archive with higher precision that
+			// we want to keep. So we skip this point.
+			if _, ok := seenTs[p.Timestamp]; ok {
```
Comment (jesusvazquez) [inline pkg/graphite/convert/whisperconverter/whisper.go:92]: @pstibrany This solves problem 2 that we discussed yesterday.

File: pkg/graphite/convert/whisperconverter/whisper.go [hunk 2]
```diff
@@ -54,53 +54,60 @@ func ReadPoints(w Archive, name string) ([]whisper.Point, error) {
...
 		}
-		// There might be duplicate timestamps in different archives. Take the
-		// higher-precision archive value since it's unaggregated.
-		if i > 0 && allPoints[i].Timestamp == allPoints[i-1].Timestamp {
-			if allPoints[i].secondsPerPoint == allPoints[i-1].secondsPerPoint {
-				return nil, fmt.Errorf("duplicate timestamp at same precision in archive %s: %d", name, allPoints[i].Timestamp)
+
+		for _, p := range archivePoints {
+			if p.Timestamp < minArchiveTs {
+				continue
```
Comment (jesusvazquez) [inline pkg/graphite/convert/whisperconverter/whisper.go:87]: @pstibrany and @ywwg This solves problem 1 where we just drop all points that are beyond the retention of the archive but still present in the ring buffer.

File: pkg/graphite/convert/whisperconverter/whisper.go [hunk 3]
```diff
@@ -54,53 +54,60 @@ func ReadPoints(w Archive, name string) ([]whisper.Point, error) {
...
 		}
-		trimmedPoints = append(trimmedPoints, allPoints[i].Point)
 	}
 
-	return trimmedPoints, nil
+	// We need to finally sort the kept points again because different archives
+	// may overlap and have older points
+	sort.Slice(keptPoints, func(i, j int) bool {
+		return keptPoints[i].Timestamp < keptPoints[j].Timestamp
+	})
```
Comment (jesusvazquez) [inline pkg/graphite/convert/whisperconverter/whisper.go:108]: With the two additions above, we now only need to sort once.

**Comment by pstibrany** [comment]
AS mentioned privately, I don't think it solves the problem that we're facing. With archives and samples like this:

Archive #0:
```
218	1663891200	2022-09-23T00:00:00Z
217	1663891320	2022-09-23T00:02:00Z
216	1663891440	2022-09-23T00:04:00Z
214	1663891560	2022-09-23T00:06:00Z
213	1663891680	2022-09-23T00:08:00Z
```

Archive #1:
```
651	1663891200	2022-09-23T00:00:00Z
427	1663891500	2022-09-23T00:05:00Z
```

I believe the end result will still be:

```
218	1663891200	2022-09-23T00:00:00Z
217	1663891320	2022-09-23T00:02:00Z
216	1663891440	2022-09-23T00:04:00Z
427	1663891500	2022-09-23T00:05:00Z *** we don't want this sample ***
214	1663891560	2022-09-23T00:06:00Z
213	1663891680	2022-09-23T00:08:00Z
```

Using archives:
```
 "Archives": [
    {
      "Offset": 40,
      "SecondsPerPoint": 60,
      "Points": 1051200 // 60s * 1051200 points = 730 days
    },
    {
      "Offset": 12614440,
      "SecondsPerPoint": 300,
      "Points": 735840 // 300s * 735840 points = 2555 days
    }
  ]
```

**Comment by jesusvazquez** [comment]
@pstibrany I've added https://github.com/grafana/mimir-proxies/pull/82/commits/07d14e00b493637ab11b7a0206537d4cb696c3f8 together with a unit test where I precisely take into account previous archives maxts to skip points in subsequent archives. Lets try again with these changes.

File: pkg/graphite/convert/whisperconverter/whisper.go [hunk 4]
```diff
@@ -54,53 +54,72 @@ func ReadPoints(w Archive, name string) ([]whisper.Point, error) {
...
+		// calculate maxArchiveTs
+		for _, p := range archivePoints {
+			// We want to track the max timestamp of the archive because we know
+			// it virtually represents now() and we wont have newer points.
+			// Then the min timestamp of the archive would be maxTs - the archive
+			// retention.
+			if p.Timestamp > maxArchiveTs {
+				maxArchiveTs = p.Timestamp
+			}
 		}
```
Comment (pstibrany) [inline pkg/graphite/convert/whisperconverter/whisper.go:79]: Why do we do this for each archive again and again?

Don't all archives have the same "now" and they only differ in their retention period?

File: pkg/graphite/convert/whisperconverter/whisper.go [hunk 5]
```diff
@@ -54,53 +54,72 @@ func ReadPoints(w Archive, name string) ([]whisper.Point, error) {
...
-		pIdx += len(archivePoints)
-	}
 
-	// Points must be in time order.
-	sort.Slice(allPoints, func(i, j int) bool {
-		return allPoints[i].Timestamp < allPoints[j].Timestamp
-	})
+		// calculate minArchiveTs
+		if maxArchiveTs-a.Retention() > maxArchiveTs { // overflow, very big retention
```
Comment (pstibrany) [inline pkg/graphite/convert/whisperconverter/whisper.go:82]:
```suggestion
		if maxArchiveTs < a.Retention() { // very big retention
```

File: pkg/graphite/convert/whisperconverter/whisper.go [hunk 6]
```diff
@@ -54,53 +54,72 @@ func ReadPoints(w Archive, name string) ([]whisper.Point, error) {
...
-	for i := 0; i < len(allPoints); i++ {
-		// Remove all points of time = 0.
-		if allPoints[i].Timestamp == 0 {
-			continue
+		// For subsequent archives
+		if lastMinTs > 0 {
+			// if we are already in the second or subsequent archive and we had
+			// some points in the prior archives, we want to skip
+			// samples in the previous archives
+			maxArchiveTs = lastMinTs
```
Comment (pstibrany) [inline pkg/graphite/convert/whisperconverter/whisper.go:93]: We update `maxArchiveTs` here, but this variable it's not used afterwards.

File: pkg/graphite/convert/whisperconverter/whisper.go [hunk 7]
```diff
@@ -54,53 +54,72 @@ func ReadPoints(w Archive, name string) ([]whisper.Point, error) {
...
 		}
-		// There might be duplicate timestamps in different archives. Take the
-		// higher-precision archive value since it's unaggregated.
-		if i > 0 && allPoints[i].Timestamp == allPoints[i-1].Timestamp {
-			if allPoints[i].secondsPerPoint == allPoints[i-1].secondsPerPoint {
-				return nil, fmt.Errorf("duplicate timestamp at same precision in archive %s: %d", name, allPoints[i].Timestamp)
+		lastMinTs = minArchiveTs
+
+		for _, p := range archivePoints {
+			if p.Timestamp < minArchiveTs {
```
Comment (pstibrany) [inline pkg/graphite/convert/whisperconverter/whisper.go:98]: Shall we check if `minArchiveTs <= p.Timestamp && p.Timestamp <= maxArchiveTs` holds here?

If we do that, we don't need to keep `seenTs`.

File: pkg/graphite/convert/whisperconverter/whisper.go [hunk 8]
```diff
@@ -54,53 +54,72 @@ func ReadPoints(w Archive, name string) ([]whisper.Point, error) {
...
+			// we want to keep. So we skip this point.
+			if _, ok := seenTs[p.Timestamp]; ok {
+				continue
 			}
-			if allPoints[i].secondsPerPoint < allPoints[i-1].secondsPerPoint {
-				trimmedPoints[len(trimmedPoints)-1] = allPoints[i].Point
+			// Skip points with time = 0
+			if p.Timestamp == 0 {
+				continue
 			}
```
Comment (pstibrany) [inline pkg/graphite/convert/whisperconverter/whisper.go:110]: I'd move this check to the of `for` loop, before we do other checks on `p.Timestamp`.

File: pkg/graphite/convert/whisperconverter/whisper.go (hunk 7 above)
Comment (pstibrany) [reply pkg/graphite/convert/whisperconverter/whisper.go:98]: Actually check for  `minArchiveTs` should be exclusive, ie. `minArchiveTs < p.Timestamp` must hold.

File: pkg/graphite/convert/whisperconverter/whisper_test.go [hunk 9]
```diff
@@ -127,6 +128,176 @@ func TestExtractWhisperPoints(t *testing.T) {
...
+				},
+				{
+					Timestamp: 1030,
+					Value:     30,
+				},
+			},
+		},
+		{
+			name:       "test retention when first archives are empty",
+			metricName: "mymetric",
```
Comment (pstibrany) [inline pkg/graphite/convert/whisperconverter/whisper_test.go:260]: This seems to be unrealistic scenario to me. If first archive (with raw data) is empty, how can there be any aggregations at all?

File: pkg/graphite/convert/whisperconverter/whisper.go (hunk 4 above)
Comment (ywwg) [reply pkg/graphite/convert/whisperconverter/whisper.go:79]: Ah, you're right.  We had seen the "offset" value in the database dump but that is a byte offset, not a timestamp offset.

File: pkg/graphite/convert/whisperconverter/whisper.go (hunk 5 above)
Comment (ywwg) [reply pkg/graphite/convert/whisperconverter/whisper.go:82]: applied manually

File: pkg/graphite/convert/whisperconverter/whisper.go (hunk 7 above)
Comment (ywwg) [reply pkg/graphite/convert/whisperconverter/whisper.go:98]: Because points are written to all archives at once, all archives will contain points from timestamps that are within the retentions. So we do need a check like this. I think we can do it more efficiently if we pre-sort the points and then do a double-index comparison

File: pkg/graphite/convert/whisperconverter/whisper_test.go (hunk 9 above)
Comment (ywwg) [reply pkg/graphite/convert/whisperconverter/whisper_test.go:260]: an end-to-end test broke, so we added this test to cover that edge case

**Comment by ywwg** [comment]
I think the logic checking to see if the point is a duplicate timestamp is redundant -- by definition we are only looking at points for timestamps where that archive is the highest resolution block.  We can be much more efficient -- start with the oldest block and just blast through the points that are within the correct bounds

**Comment by ywwg** [comment]
a much faster algo is:

* determine maxTs by finding the highest ts in the first archive
* calculate ts bounds for each archive
* loop through, starting at lowest resolution block.
  * sort its points
  * only add points to Kept that are within the bounds
 

done

File: pkg/graphite/convert/whisperconverter/whisper.go [hunk 10]
```diff
@@ -40,67 +41,85 @@ type Archive interface {
...
+		endIdx := len(points) - 1
+		for j, p := range points {
+			if p.Timestamp == 0 {
+				continue
 			}
-			if allPoints[i].secondsPerPoint < allPoints[i-1].secondsPerPoint {
-				trimmedPoints[len(trimmedPoints)-1] = allPoints[i].Point
+			// Don't include any points in this archive that are past the retention
+			// period.
+			if p.Timestamp < minArchiveTs {
```
Comment (pstibrany) [inline pkg/graphite/convert/whisperconverter/whisper.go:102]: Nit: I think it should be

```suggestion
			if p.Timestamp <= minArchiveTs {
```

Point with Timestamp equal to `minArchiveTs` is already outside of retention.

File: pkg/graphite/convert/whisperconverter/whisper.go [hunk 11]
```diff
@@ -40,67 +41,85 @@ type Archive interface {
...
+				continue
 			}
-			// If the previous point is higher precision, just continue.
-			continue
+			// Don't include any points in this archive that were covered in a higher
+			// resolution archive.
+			if p.Timestamp >= lastMinTs {
```
Comment (pstibrany) [inline pkg/graphite/convert/whisperconverter/whisper.go:107]: Similarly:

```suggestion
			if p.Timestamp > lastMinTs {
```

File: pkg/graphite/convert/whisperconverter/whisper_test.go (hunk 9 above)
Comment (pstibrany) [reply pkg/graphite/convert/whisperconverter/whisper_test.go:256]: :+1:

**Review by pstibrany — APPROVED** [review]
Thanks, this looks good.

I think we have tiny mistake in bounds computation (archive with 10 points, with max timestamp 100 and resolution 1 should NOT include point 90).

File: pkg/graphite/convert/whisperconverter/whisper.go (hunk 11 above)
Comment (ywwg) [reply pkg/graphite/convert/whisperconverter/whisper.go:107]: ahh I was wondering if I'd got that switched.  thanks

**Comment by ywwg** [comment]
thanks for the bounds-check fix!

**Comment by ywwg** [comment]
explaining the new tests I had to write, this is a valid archive:

```
Meta data:
  aggregation method: sum
  max retention: 2012232704

Archive 0 info:
  offset: 52
  seconds per point: 1
  points: 86400
  retention: 86400
  size: 1036800

Archive 1 info:
  offset: 1036852
  seconds per point: 3600
  points: 840
  retention: 3024000
  size: 10080

Archive 2 info:
  offset: 1046932
  seconds per point: 86400
  points: 73000
  retention: 6307200000
  size: 876000

Archive 0 data:
Archive 1 data:
Archive 2 data:
0: 1517443200,          1
1: 1517529600,          1
2: 1517616000,          1
```

**Review by pstibrany — APPROVED** [review]
Code makes sense to me. Thanks.
